yoda/types.h

```diff
@@ -46,10 +48,83 @@ struct AllowOverwriteOnAdd {
   constexpr static bool allow_overwrite_on_add = true;
 };
 
-// Entry key type extractor.
+// Entry key type extractor, getter and setter.
+// Supports both `.key` data member and `.key() / .set_key()` methods.
 template <typename T_ENTRY>
-using ENTRY_KEY_TYPE = typename std::remove_cv<
-    typename std::remove_reference<decltype(std::declval<T_ENTRY>().key())>::type>::type;
+constexpr bool HasKeyFunction(char) {
+  return false;
+}
+
+template <typename T_ENTRY>
+constexpr auto HasKeyFunction(int) -> decltype(std::declval<T_ENTRY>().key(), bool()) {
+  return true;
+}
+
+template <typename T_ENTRY, bool HAS_KEY_FUNCTION>
+struct KEY_ACCESSOR_IMPL {};
+
+template <typename T_ENTRY>
+struct KEY_ACCESSOR_IMPL<T_ENTRY, false> {
+  typedef decltype(std::declval<T_ENTRY>().key) T_KEY;
+  static T_KEY GetKey(const T_ENTRY& entry) { return entry.key; }
```

Comment:
Just a question - you chose to return by value because most of the key types are simple or use rvalue semantics, right?